test/catch2_test_warp_reduce.cu

```diff
@@ -0,0 +1,598 @@
+{
+  REQUIRE_APPROX_EQ(expected_data, test_results);
+}
+
+/**
+ * @brief Compares the results returned from system under test against the expected results.
+ */
+template <typename T,
+          typename ::cuda::std::enable_if<!::cuda::std::is_floating_point<T>::value, int>::type = 0>
+void verify_results(const thrust::host_vector<T> &expected_data,
+                    const thrust::device_vector<T> &test_results)
+{
+  REQUIRE(expected_data == test_results);
+}
+
+enum class reduce_mode
+{
+  all,
+  partial,
+  head_flags,
+  tail_flags,
+};
+
+template <typename InputItT, typename FlagInputItT, typename ReductionOp, typename ResultOutItT>
+void compute_host_reference(reduce_mode mode,
+                            InputItT h_in,
+                            FlagInputItT h_flags,
+                            int warps,
+                            int logical_warp_threads,
+                            int valid_warp_threads,
+                            ReductionOp reduction_op,
+                            ResultOutItT h_data_out)
+{
+  // Accumulate segments (lane 0 of each warp is implicitly a segment head)
+  for (int warp = 0; warp < warps; ++warp)
+  {
+    int warp_offset = warp * logical_warp_threads;
+    int item_offset = warp_offset + valid_warp_threads - 1;
+
+    // Last item in warp
+    auto head_aggregate = h_in[item_offset];
+    auto tail_aggregate = h_in[item_offset];
+
+    if (mode != reduce_mode::tail_flags && h_flags[item_offset])
+    {
+      h_data_out[item_offset] = head_aggregate;
+    }
+    item_offset--;
+
+    // Work backwards
+    while (item_offset >= warp_offset)
+    {
+      if (h_flags[item_offset + 1])
+      {
+        head_aggregate = h_in[item_offset];
+      }
+      else
+      {
+        head_aggregate = reduction_op(head_aggregate, h_in[item_offset]);
+      }
+
+      if (h_flags[item_offset])
+      {
+        if (mode == reduce_mode::head_flags)
+        {
+          h_data_out[item_offset] = head_aggregate;
+        }
+        else if (mode == reduce_mode::tail_flags)
+        {
+          h_data_out[item_offset + 1] = tail_aggregate;
+          tail_aggregate              = h_in[item_offset];
+        }
+      }
+      else
+      {
+        tail_aggregate = reduction_op(tail_aggregate, h_in[item_offset]);
+      }
+
+      item_offset--;
+    }
+
+    // Record last segment aggregate
+    if (mode == reduce_mode::tail_flags)
+    {
+      h_data_out[warp_offset] = tail_aggregate;
+    }
+    else
+    {
+      h_data_out[warp_offset] = head_aggregate;
+    }
+  }
+}
+
+// List of types to test
+using custom_t =
+  c2h::custom_type_t<c2h::accumulateable_t, c2h::equal_comparable_t, c2h::less_comparable_t>;
+using types = c2h::type_list<std::uint8_t,
+                             std::uint16_t,
+                             std::int32_t,
+                             std::int64_t,
+                             custom_t,
+                             ulonglong4,
```

Comment:
Do you think there might be a value in extracting test of vector types out of the full configuration space?